### PR TITLE
feat: implement backoff jitter retry mechanism

### PR DESCRIPTION
**Pull request by StarpTech**
## Motivation and Context

Implements backoff jitter retry for query operations
https://cosmo-docs.wundergraph.com/~/changes/VKKWOoLA1Kg0m70D5CDW/router/traffic-shaping

Introduce a breaking change where delays, and timeouts are parsed as Go duration (10s, 100ms) instead of suffixing all environment variables and config properties with _seconds)

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: router/core/factoryresolver.go [hunk 1]
```diff
@@ -66,138 +60,13 @@ func NewDefaultFactoryResolver(transportFactory ApiTransportFactory, baseTranspo
...
-		// if the data source uses a custom proxy, create a dedicated client
-		if dataSourceUsesHTTPProxy(ds) {
-			_, found := config.LookupStringVariable(cfg.HttpProxyUrl)
-			return found
-		}
-	}
-	return false
-}
-
-// customTLSTransport returns a TLS *http.Transport with the given key and certificates loaded
```
Comment (StarpTech) [inline router/core/factoryresolver.go:89]: Legacy code from WunderGraph. No code was actually used.

File: router/internal/retrytransport/retry_transport.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		resp, err = rt.RoundTripper.RoundTrip(req)
+
+	}
+
+	return resp, err
+}
+
+func IsRetryableError(err error, resp *http.Response) bool {
+
+	if resp != nil {
```
Comment (jensneuse) [inline router/internal/retrytransport/retry_transport.go:96]: I'd put err check above res check.

File: router/internal/retrytransport/retry_transport.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+			zap.Int("retry", retries),
+			zap.String("url", req.URL.String()),
+			zap.Duration("sleep", sleepDuration),
+		)
+
+		// Wait for the specified backoff period
+		time.Sleep(sleepDuration)
+
+		// Retry the request
+		resp, err = rt.RoundTripper.RoundTrip(req)
```
Comment (jensneuse) [inline router/internal/retrytransport/retry_transport.go:87]: I'd add below pseudocode:
If err is nil and res status code is 200, return it.
No need to evaluate further and it makes the flow clear that once we have a valid response, we return it immediately.

File: router/internal/retrytransport/retry_transport.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	return &RetryHTTPTransport{
+		RoundTripper: roundTripper,
+		RetryOptions: retryOptions,
+		Logger:       logger,
+	}
+}
+
+func (rt *RetryHTTPTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+
+	resp, err := rt.RoundTripper.RoundTrip(req)
```
Comment (jensneuse) [inline router/internal/retrytransport/retry_transport.go:60]: I'd add pseudocode after this:
If err is nil and resp status code is 200, return immediately.
No need to create a backoff or do further evaluations if we should retry.

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: router/internal/retrytransport/retry_transport.go (hunk 3 above)
Comment (StarpTech) [reply router/internal/retrytransport/retry_transport.go:87]: :+1:

**Review by jensneuse — APPROVED** [review]

**Pull request merged**